### PR TITLE
feat: Respect specified target framework only if this results in a valid version

### LINKED ISSUE #3038: Upgrading from 4.0.6 to 4.2.0 analysis no longer works for project
**Describe the bug**

Attempting to upgrade from 4.0.6 to 4.2.0, analysis is failing for one of our test projects, causing an error.

Specifying `--dev-mode` doesn't appear to add any additional details.

### 4.0.6

```console
Running mutation tests for '../src/Polly/Polly.csproj'. Test Project: 'Polly.Specs/Polly.Specs.csproj'

   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 4.0.6

A new version of Stryker.NET (4.2.0) is available. Please consider upgrading 
using `dotnet tool update -g dotnet-stryker`

[09:02:00 INF] Analysis starting.
[09:02:06 WRN] Could not find a project analysis for the chosen target framework net8.0. 
The available target frameworks are: net462,net472,net6.0,netstandard2.0. 
The first available framework will be selected, which is net462.
[09:02:06 INF] Found project D:\a\Polly\Polly\src\Polly\Polly.csproj to mutate.
[09:02:06 INF] Analysis complete.
[09:02:06 INF] Building test project D:\a\Polly\Polly\test\Polly.Specs\Polly.Specs.csproj (1/1)
[09:03:18 INF] Number of tests found: 1986 for project D:\a\Polly\Polly\src\Polly\Polly.csproj. Initial test run started.
```

### 4.2.0

```console
Running mutation tests for '../src/Polly/Polly.csproj'. Test Project: 'Polly.Specs/Polly.Specs.csproj'

   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 4.2.0

[12:05:57 INF] Analysis starting.
[12:05:57 INF] Analyzing 1 test project(s).
[12:06:11 INF] Analysis complete.
[12:06:11 INF] Time Elapsed 00:00:13.4378512
Stryker.NET failed to mutate your project. For more information see the logs 
below:


No project references found. Please add a project reference to your test project
and retry.
```

**Logs**

See the `mutation-report` artifact [here](https://github.com/App-vNext/Polly/actions/runs/10922104543?pr=2293).

**Expected behavior**

Mutation tests work.

**Desktop (please complete the following information):**
 - OS: Windows
 - Type of project: framework and core
 - Framework Version: `net462,net472,net6.0,netstandard2.0`
 - Stryker Version: `4.2.0`

**Additional context**

Failing upgrade: https://github.com/App-vNext/Polly/pull/2293

### PR DESCRIPTION
**Pull request by dupdob**
This PR reverts the usage of target framework during project analysis (by Buildalyzer). That is the target framework is no longer specified when calling Buildalyzer (see below why).
This how Stryker pick a target framework version now:

1. All projects are analyzed without specifying a target framework
2. For test projects
    1.  If a target version is specified, and is present and valid in the analysis results, Stryker picks this framework version for the given test project
    2. otherwise, Stryker picks the first net version, and if none is available, picks the first version
3. for tested projects
    1.  Stryker picks the exact version used by the test project (that references this project)

The overall logic can summarized as: Stryker uses test project with the configured target framework when available, and relies on MsBuild to identify proper dependencies for the other projects. This leads to more consistent results.

> **Note:** when we provide a target framework to Buildalyzer/Msbuild, it will analyze/build an assembly for the target framework, even if it is not explicitely listed in the _csproj_ file. As such, Buildalyzer does report a success. This could lead to situation where the analysis succeed but mutation can not happen as the requested version is not actually built. Leading to surprising failure. On the other hand, we could have an option in Stryker that would force the project(s) to be built against the specified framework, but I am not sure there is a need for that.

~~Build analysis may fail when providing a target framework version that is not listed as a valid target by the project.
This PR ensures analysis retry does not ask for a specific framework to work around this issue.
Then, the legacy framework versions selection takes care of the rest.~~

~~May~~ fix #3038

### CODE REVIEW

**Comment by rouke-broersma** [comment]
> For test projects
If a target version is specified, and is present and valid in the analysis results, Stryker picks this framework version for the given test project

I'm really unsure if this is what users will usually intend when setting the target framework. I would expect that the project under test will most likely have multiple target frameworks and that the unit test project does not necessarily need to have multiple target frameworks.

**Comment by dupdob** [comment]
> > For test projects
> > If a target version is specified, and is present and valid in the analysis results, Stryker picks this framework version for the given test project
> 
> I'm really unsure if this is what users will usually intend when setting the target framework. I would expect that the project under test will most likely have multiple target frameworks and that the unit test project does not necessarily need to have multiple target frameworks.
Actually, you need multi targeting test projects if you intend to test your various targets. If you do that, it implies you have no variation across target framework versions, hence you do not need multi targeting.
In any case, if you only have single targeting test projects, you do not need this option.
> 
> What happens when a project under test is referenced by multiple unit test projects?
Stryker picks the target framework version used by the first detected test projects and keeps it for all other test projects.

Here is what happens depending on Sryker version
## Old versions (<4.1)

|mode| test project | picked | target project | picked |comment]
|----|-------------|-------|---------------|---------|---------|
|no option| net3.0, netfwk4.62 |net3.0      |    netfwk4.61, net2.0| netfwk4.61 | **testing fails, can't use net3.0 can't depend on netfwk4.61**|
|no option| net3.0, net2.0 |net3.0   |  net2.0, net3.0| net2.0 | works but does not use expected fmwk for target project|
|targetframework=net2.0| net3.0, netfwk4.62 |net3.0  |    netfwk4.61, net2.0| net2.0 | works|
|targetframework=netfwk4.61| net3.0, netfwk4.62 |net3.0      |    netfwk4.61, net2.0| netfwk4.61 | **testing fails, can't use net3.0 can't depend on netfwk4.61**|

The problem with Stryker pre 4.1 is that it picks version at random when it does not find the expected version with no guaranteed compatibility.

## Current version (4.1+)

|mode| test project | picked | target project | picked |comment]
|----|-------------|-------|---------------|---------|---------|
|no option| net3.0, netfwk4.62 |net 3.0      |    netfwk4.61, net2.0| net2.0 |works, Stryker detected that it must use net2.0|
|no option| net3.0, net2.0 |net3.0   |  net2.0, net3.0| net3.0 | works, Stryker detected that it must use net3.0|
|targetframework=net2.0| net3.0, netfwk4.62 |net3.0  |    netfwk4.61, net2.0| net2.0 | works|
|targetframework=net2.0| net3.0, netfwk4.62 |net2.0      |    netfwk4.61, net2.0| net2.0 | **fails, as there is no net2.0 target for test project**|
|targetframework=net3.0| net3.0, netfwk4.62 |net3.0      |    netfwk4.61, net2.0| net3.0 |**fails, as there is no net3.0 target for target project**|
|targetframework=netfwk4.61| net3.0, netfwk4.62 |netfwk4.61      |    netfwk4.61, net2.0| netfwk4.61.0 | **testing fails, can't use net3.0 can't depend on netfwk4.61**|

Current versions work better when no target is specified as it correctly resolves dependencies, but if a non available target is specified, it will result in errors because Buildalyzer generates phantom targets that are not actually build

## This PR
|mode| test project | picked | target project | picked |comment]
|----|-------------|-------|---------------|---------|---------|
|no option| net3.0, netfwk4.62 |net 3.0      |    netfwk4.61, net2.0| net2.0 |works, Stryker detected that it must use net2.0|
|no option| net3.0, net2.0 |net3.0   |  net2.0, net3.0| net3.0 | works, Stryker detected that it must use net3.0|
|targetframework=net2.0| net3.0, netfwk4.62 |net3.0  |  netfwk4.61, net2.0| net2.0 | works|
|targetframework=net3.0| net3.0, netfwk4.62 |net3.0  |  netfwk4.61, net2.0| net2.0 |works|
|targetframework=netfwk4.61| net3.0, netfwk4.62 |net3.0      |    netfwk4.61, net2.0| net2.0 | works|

Basically this PR fixes situations where the current versions do fail. This is also an opportunity to clarify the selection logic.
We can also have a discussion regarding what could be the best logic. Here I am just aligning with expectations in issue #3038, which makes sense to me.

**Comment by dupdob** [comment]
Having unit tests would be great to track and document this. I switch this PR back to draft and see if those could be added

**Comment by rouke-broersma** [comment]
Thanks for this!
